Source/CDRSpec.m

```diff
@@ -18,43 +49,96 @@ void afterEach(CDRSpecBlock block) {
     [CDR_currentSpec.currentGroup addAfter:block];
 }
 
-#define with_stack_address(b) \
-    ((b.stackAddress = CDRCallerStackAddress()), b)
+void ensureCurrentSpecExists(NSString *functionName) {
+    if (!CDR_validateSpecs) {
+        return;
+    }
 
-CDRExampleGroup * describe(NSString *text, CDRSpecBlock block) {
-    CDRExampleGroup *group = nil;
-    if (block) {
-        CDRExampleGroup *parentGroup = CDR_currentSpec.currentGroup;
-        group = [CDRExampleGroup groupWithText:text];
-        [parentGroup add:group];
-        CDR_currentSpec.currentGroup = group;
-
-        @try {
-            block();
-        }
-        @catch (CDRSpecFailure *failure) {
-            NSString *reason = [NSString stringWithFormat:@"Caught a spec failure before the specs began to run. Did you forget to put your assertion into an `it` block?. The failure was:\n%@", failure];
-            [[NSException exceptionWithName:NSInternalInconsistencyException reason:reason userInfo:nil] raise];
-        }
+    if (CDR_currentSpec == nil) {
+        NSString * reason = [NSString stringWithFormat:@"%@() was invoked outside of a spec. It may only be called when a spec has been defined with SPEC_BEGIN and SPEC_END macros.", functionName];
+        [[NSException exceptionWithName:NSInternalInconsistencyException
+                                 reason:reason
+                               userInfo:nil] raise];
+    }
+}
 
-        if ([group.examples count] == 0) {
-            block = placeholderPendingTestBlock;
-            block();
+void ensureTestsAreNotYetRunning(NSString *functionName) {
+    if (!CDR_validateSpecs) {
+        return;
+    }
+
+    switch (CDRCurrentState()) {
+        case CedarRunStateNotYetStarted:
+            // exceedingly unlikely
+            break;
+        case CedarRunStateBuildingTree:
+            // should hit this 100% of the time
+            break;
+        case CedarRunStateRunningTests: {
+            // someone done goofed
+            NSString * reason = [NSString stringWithFormat:@"%@() was invoked during a test run. Make sure your '%@()' is not inside of an it() block.", functionName, functionName];
+            [[NSException exceptionWithName:NSInternalInconsistencyException
+                                     reason:reason
+                                   userInfo:nil] raise];
+            break;
         }
-        CDR_currentSpec.currentGroup = parentGroup;
-    } else {
-        group = describe(text, placeholderPendingTestBlock);
+        case CedarRunStateFinished:
+            // someone ... REALLY done goofed
```

Comment:
Absolutely, although it's pretty tough to trigger this in a realistic test, since you'd need to call `it()` and assert that an exception is raised sometime after `main()` and before Cedar starts to run.